zcash_client_sqlite/src/wallet.rs

```diff
@@ -816,29 +822,39 @@ pub(crate) fn get_transparent_balances<P: consensus::Parameters>(
 }
 
 /// Inserts information about a scanned block into the database.
-pub(crate) fn insert_block(
+pub(crate) fn put_block(
     conn: &rusqlite::Connection,
     block_height: BlockHeight,
     block_hash: BlockHash,
     block_time: u32,
     sapling_commitment_tree_size: Option<u64>,
 ) -> Result<(), SqliteClientError> {
-    let mut stmt_insert_block = conn.prepare_cached(
+    let mut stmt_upsert_block = conn.prepare_cached(
         "INSERT INTO blocks (
             height,
             hash,
             time,
             sapling_commitment_tree_size,
             sapling_tree
         )
-        VALUES (?, ?, ?, ?, x'00')",
+        VALUES (
+            :height,
+            :hash,
+            :block_time,
+            :sapling_commitment_tree_size,
+            x'00'
+        )
+        ON CONFLICT (height) DO UPDATE
+        SET hash = :hash,
+            time = :block_time,
+            sapling_commitment_tree_size = :sapling_commitment_tree_size",
```

Comment:
Hm. Automatically doing a reorg as a side-effect of `put_block` seems like a really dangerous thing to do. `ON CONFLICT (height, hash)` is possibly the right thing; the objective here is to ensure that if a block is scanned twice, the operation is idempotent; if you attempt to put a block with a different hash at the same height, that _should_ be an error given the existing wallet data structure.

Reply:
Oops, except, `ON CONFLICT (height, hash)` also doesn't work, because the `ON CONFLICT` clause must match the `PRIMARY KEY` or a `UNIQUE` constraint on the table.

Reply:
In light of that, this is going to have to be a transactional query-then-write.

Reply:
Fixed in c4e11e684d81ec2bfacbce9d52c2854308176385